### ISSUE TITLE
Plotting Spikes,weights,Performance

### ISSUE THREAD

**Issue by tehreemnaqvi** [HiroshiARAKI/snnlibpy#1]
Hi,
I just started to follow your repository.You have done very well.
So, I wanted to plot my input spikes,weight and model performance but after training my plot didn't show.
I am confused how to solve this issue Can you tell me, how can I fix this issue?
My code is like this:
for epoch in range( n_epochs ):
    if epoch % progress_interval == 0:
        print( "Progress: %d / %d (%.4f seconds)" % (epoch, n_epochs, t() - start) )
        start = t()

    train_dataloader = torch.utils.data.DataLoader(
        train_dataset, batch_size=1, shuffle=True)

    for step, batch in enumerate( train_dataloader ):
        # Get next input sample.

        inputs = {"X": batch["encoded_image"].view( time, 1, 1, 28, 28 )}
        if gpu:
            inputs = {k: v.cuda() for k, v in inputs.items()}
        label = batch["label"]

        # Run the network on the input.
        network.run( inputs=inputs, time=time, input_time_dim=1 )

        # Optionally plot various simulation information.
        if plot:
            image = batch["image"].view( 28, 28 )

            inpt = inputs["X"].view( time, 784 ).sum( 0 ).view( 28, 28 )
            weights1 = conv_conn.w
            _spikes = {
                "X": spikes["X"].get( "s" ).view( time, -1 ),
                "Y": spikes["Y"].get( "s" ).view( time, -1 ),
            }
            _voltages = {"Y": voltages["Y"].get( "v" ).view( time, -1 )}

            inpt_axes, inpt_ims = plot_input(
                image, inpt, label=label, axes=inpt_axes, ims=inpt_ims
            )
            spike_ims, spike_axes = plot_spikes( _spikes, ims=spike_ims, axes=spike_axes )
            weights1_im = plot_conv2d_weights( weights1, im=weights1_im )
            voltage_ims, voltage_axes = plot_voltages(
                _voltages, ims=voltage_ims, axes=voltage_axes
            )

            #plt.ioff()
            #plt.show()

        plt.ioff()
        # plt.pause( 1 )
        plt.show()

        network.reset_state_variables()  # Reset state variables.

print( "Progress: %d / %d (%.4f seconds)\n" % (n_epochs, n_epochs, t() - start) )
print( "Training complete.\n" )

**Comment by HiroshiARAKI** [comment]
Hi.

I have no idea about your problem completely.
So, before I answer your question, I have some questions for you.

1. Do you know what error message said? If you know that please tell me.
2. Do you use GPU computing??
3. Didn't all plotting function work? (e.g. "Only Plotting spikes work properly," "only plotting weights doesn't work well, " or something like that) 
4. If your network is before training, all plotting functions work well??

that's all my questions.
If these questions' answers are crystal-clear, I may solve your problem.

**Comment by tehreemnaqvi** [comment]
Hi, Actually  i am trying to plot these spikes after training.I am also using GPU and once training is complete, none of the spikes are generated.I didn't get any error message after training is completed.

**Comment by HiroshiARAKI** [comment]
Hmm... I've not known well about that problem yet, but...
If you put on "plt.close()" after "plt.show()" or "plt.savefig()", how does your program work?

**Comment by tehreemnaqvi** [comment]
Hi,This is my code.
Can you see what's the problem in this code?

    for epoch in range( args.n_epochs ):
        # Create a dataloader to iterate over dataset.
        dataloader = DataLoader(
            dataset,
            batch_size=args.batch_size,
            shuffle=True,
            num_workers=args.n_workers,
            pin_memory=args.gpu,
        )

        for step, batch in enumerate( tqdm( dataloader ) ):
            # Prep next input batch.
            inpts = {"I": batch["encoded_image"]}
            if args.gpu:
                inpts = {k: v.cuda() for k, v in inpts.items()}

            # Run the network on the input.
            network.run( inputs=inpts, time=args.time )

            # Plot simulation data.
            if args.plot:
                spikes = {}
                for name, monitor in network.monitors.items():
                    spikes[name] = monitor.get( "s" )[:, 0].view( args.time, -1 )

                spike_ims, spike_axes = plot_spikes(
                    spikes, ims=spike_ims, axes=spike_axes
                )

                conv1_weights_im = plot_conv2d_weights(
                    conv1_connection.w, im=conv1_weights_im, wmin=-1.0, wmax=1.0
                )
                conv2_weights_im = plot_conv2d_weights(
                    conv2_connection.w, im=conv2_weights_im, wmin=-1.0, wmax=1.0
                )
                dense_weights_im = plot_weights(
                    dense_connection.w, im=dense_weights_im, wmin=-1.0, wmax=1.0
                )
                output_weights_im = plot_weights(
                    output_connection.w, im=output_weights_im, wmin=-1.0, wmax=1.0
                )
                #plt.ioff()
               
                plt.show()
                plt.close()

            network.reset_state_variables()

**Comment by tehreemnaqvi** [comment]
When i run the program after calling plt.close().I had the same output and did not get any plot.
This is the output after training.
![output](https://user-images.githubusercontent.com/38427571/81500217-9f043d80-930b-11ea-8b49-38adf7abe727.PNG)

**Comment by HiroshiARAKI** [comment]
Hi, 
I can't find where the essential problem is.

How about that you stop giving 'im (AxesImage object)'  as arguments and change to call "plt.show" or "plt.savefig()" after each plotting functions.
So, what happens if you modify a code as below.

```
plot_conv2d_weights(conv1_connection.w,  wmin=-1.0, wmax=1.0)
plt.savefig('****.png')
plt.close()

plot_conv2d_weights(conv2_connection.w, wmin=-1.0, wmax=1.0)
plt.savefig('****.png')
plt.close()

plot_weights(dense_connection.w, wmin=-1.0, wmax=1.0)
plt.savefig('****.png')
plt.close()

plot_weights(output_connection.w, wmin=-1.0, wmax=1.0)
plt.savefig('****.png')
plt.close()
```

If it does not work well, you try to do with one plotting function.
For example, only
```
plot_weights(output_connection.w, wmin=-1.0, wmax=1.0)
plt.savefig('****.png')
plt.close()
```
this code.

Maybe, this problem is about matplotlib, not bindsnet.

**Comment by tehreemnaqvi** [comment]
Thank you very much.
Now, i will try this method.

**Comment by HiroshiARAKI** [comment]
And, check my plotting code out.

[https://github.com/HiroshiARAKI/snnlibpy/blob/master/wbn/snnlib.py#L870](https://github.com/HiroshiARAKI/snnlibpy/blob/master/wbn/snnlib.py#L870)

This code using bindsnet's plotting functions works properly on my environment.

**Comment by tehreemnaqvi** [comment]
Thank you so much.
I will refer this.

**Comment by tehreemnaqvi** [comment]
Your code don't run on Windows?Right?

**Comment by tehreemnaqvi** [comment]
I tried the way you told me about plotting but again did not get any plot.

**Comment by tehreemnaqvi** [comment]
Now, this is my code:


    for epoch in range( args.n_epochs ):
        # Create a dataloader to iterate over dataset.
        dataloader = DataLoader(
            dataset,
            batch_size=args.batch_size,
            shuffle=True,
            num_workers=args.n_workers,
            pin_memory=args.gpu,
        )

        for step, batch in enumerate( tqdm( dataloader ) ):
            # Prep next input batch.
            inpts = {"I": batch["encoded_image"]}
            if args.gpu:
                inpts = {k: v.cuda() for k, v in inpts.items()}

            # Run the network on the input.
            network.run( inputs=inpts, time=args.time )

            # Plot simulation data.
            if args.plot:
                spikes = {}
                for name, monitor in network.monitors.items():
                    spikes[name] = monitor.get( "s" )[:, 0].view(
args.time, -1 )

                spike_ims, spike_axes = plot_spikes(
                    spikes, ims=spike_ims, axes=spike_axes
                )

                #conv1_weights_im = plot_conv2d_weights(
                    #conv1_connection.w, im=conv1_weights_im,
wmin=-1.0, wmax=1.0
                #)
                plot_conv2d_weights( conv1_connection.w, wmin=-1.0, wmax=1.0 )
                plt.savefig( '1.png' )
                plt.close()

                plot_conv2d_weights( conv2_connection.w, wmin=-1.0, wmax=1.0 )
                plt.savefig( '2.png' )
                plt.close()

                plot_weights( dense_connection.w, wmin=-1.0, wmax=1.0 )
                plt.savefig( '3.png' )
                plt.close()

                plot_weights( output_connection.w, wmin=-1.0, wmax=1.0 )
                plt.savefig( '4.png' )

                plt.show()
                plt.close()

                #conv2_weights_im = plot_conv2d_weights(
                    #conv2_connection.w, im=conv2_weights_im,
wmin=-1.0, wmax=1.0
                #)
                #dense_weights_im = plot_weights(
                    #dense_connection.w, im=dense_weights_im,
wmin=-1.0, wmax=1.0
                #)
               # output_weights_im = plot_weights(
                    #output_connection.w, im=output_weights_im,
wmin=-1.0, wmax=1.0
                #)
                #plt.ioff()

                #plt.draw()
                #plt.pause(0.001)
                #plt.close()

        network.reset_state_variables()

**Comment by tehreemnaqvi** [comment]
Thank you very much for your help.
Problem solved by plt.savefig().

**Comment by HiroshiARAKI** [comment]
Hi, you're welcome. Im glad to solve your problem.
